### PR TITLE
Ddf 1733: Expose catalog endpoints

### PR DESCRIPTION
**Pull request by gordocanchola**
@clockard @djblue @vinamartin

### CODE REVIEW

File: catalog/spatial/csw/spatial-csw-endpoint/pom.xml [hunk 1]
```diff
@@ -146,22 +146,22 @@
                                         <limit>
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.75</minimum>
+                                            <minimum>0.25</minimum>
```
Comment (clockard) [inline catalog/spatial/csw/spatial-csw-endpoint/pom.xml]: Why the drastic drop in coverage?

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 2]
```diff
@@ -44,4 +44,15 @@
         <argument ref="systemInfo"/>
     </bean>
 
+    <bean id="openSearchCatalogEndpoint" class="ddf.catalog.endpoint.impl.CatalogEndpointImpl">
+        <property name="url" value="${org.codice.ddf.system.protocol}${org.codice.ddf.system.hostname}:${org.codice.ddf.system.port}${org.codice.ddf.system.rootContext}/catalog/query"/>
+        <property name="urlBindingName" value="openSearchUrl"/>
+        <property name="id" value="registry.federation.method.openSearch"/>
```
Comment (clockard) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: The id values for the endpoints are very registry focused. I am thinking now that we should change that. The id could really be about anything we wanted it to be as long as it is unique amongst the endpoints on a system. Maybe just something like ddf.catalog.endpoint.opensearch

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 3]
```diff
@@ -44,4 +44,15 @@
...
+        <property name="name" value="Open Search Endpoint"/>
+        <property name="version" value="1.0.0"/>
+        <property name="description" value="The endpoint used for open search. Fix this."/>
```
Comment (clockard) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Remove "Fix this"

**Comment by vinamartin** [comment]
I'll hero this.
![hero]
(http://zuzmoka.hu/wp-content/uploads/2015/01/batman.gif)

**Comment by gordocanchola** [comment]
Modified the blueprint files to have the updated id and removed the 'Fix this'

Reverted the coverage numbers and the build didn't complain.

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 3 above)
Comment (gordocanchola) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Removed "Fix this"

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 2 above)
Comment (gordocanchola) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Updated the id

File: catalog/spatial/csw/spatial-csw-endpoint/pom.xml (hunk 1 above)
Comment (gordocanchola) [inline catalog/spatial/csw/spatial-csw-endpoint/pom.xml]: Probably just dropped them to get the build to pass. Reverted the coverage changes and the build didn't complain.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
+/**
```
Comment (vinamartin) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: There's a few methods in this class where I can't find any usages. 
Should we get rid of them or are you planning on using them in the future?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 5]
```diff
@@ -147,8 +153,8 @@ public void tearDown() throws Exception {
             }
             metacardsToDelete.clear();
         }
-        urlResourceReaderConfigurator.setUrlResourceReaderRootDirs(
-                new String[] {DEFAULT_URL_RESOURCE_READER_ROOT_RESOURCE_DIRS});
+        urlResourceReaderConfigurator.setUrlResourceReaderRootDirs(new String[] {
```
Comment (vinamartin) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Is there a reason why doing this won't work? 
`urlResourceReaderConfigurator.setUrlResourceReaderRootDirs(DEFAULT_URL_RESOURCE_READER_ROOT_RESOURCE_DIRS);`

**Comment by vinamartin** [comment]
Got a successful build. :+1: After all other comments are addressed.

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/endpoint/CatalogEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.endpoint;
+
+import java.util.Map;
+
+public interface CatalogEndpoint {
+    String BINDING_TYPE_KEY = "bindingType";
```
Comment (oscaritoro) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/endpoint/CatalogEndpoint.java]: should these key Strings also be "public static final"?

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java (hunk 4 above)
Comment (oscaritoro) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: Did you do a final reformat of this file? Doing so on my intelliJ actually moved some stuff around.

Comment (oscaritoro) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: Shouldn't you be using proper javadoc? The one liner comments don't do much more than just echo the name.

File: catalog/opensearch/catalog-opensearch-endpoint/pom.xml [hunk 7]
```diff
@@ -156,7 +156,8 @@
               ddf.catalog.impl,
```
Comment (oscaritoro) [inline catalog/opensearch/catalog-opensearch-endpoint/pom.xml]: Should we be reformatting this file since we're touching it?

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 8]
```diff
@@ -44,4 +44,15 @@
         <argument ref="systemInfo"/>
     </bean>
 
+    <bean id="ddf.catalog.endpoint.opensearch" class="ddf.catalog.endpoint.impl.CatalogEndpointImpl">
+        <property name="url" value="${org.codice.ddf.system.protocol}${org.codice.ddf.system.hostname}:${org.codice.ddf.system.port}${org.codice.ddf.system.rootContext}/catalog/query"/>
```
Comment (oscaritoro) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Should we be reformatting this file since we're touching it?

File: catalog/registry/catalog-registry-cswinputtransformer/pom.xml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ **/
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>ddf.registry</groupId>
+        <artifactId>registry</artifactId>
+        <version>2.9.0-SNAPSHOT</version>
+    </parent>
+    <groupId>ddf.registry</groupId>
```
Comment (oscaritoro) [inline catalog/registry/catalog-registry-cswinputtransformer/pom.xml]: should we be reusing the groupId here? (shows up as warning, but might be ok)

File: catalog/registry/pom.xml [hunk 10]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (oscaritoro) [inline catalog/registry/pom.xml]: Should we be reformatting this file since we just created it?

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java (hunk 4 above)
Comment (oscaritoro) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: Should we be creating a Unit test for this new class implementation?

File: catalog/spatial/csw/spatial-csw-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 11]
```diff
@@ -116,6 +116,17 @@
         <argument ref="metacardTransformerManager"/>
```
Comment (oscaritoro) [inline catalog/spatial/csw/spatial-csw-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Should we be reformatting this file since we touched it?

File: catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 12]
```diff
@@ -76,6 +76,17 @@
         <argument ref="schemaCache"/>
```
Comment (oscaritoro) [inline catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Should we be reformatting this file since we just touched it?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 13]
```diff
@@ -169,11 +175,18 @@ public void tearDown() throws Exception {
     public void testFederatedQueryByWildCardSearchPhrase() throws Exception {
         String queryUrl = OPENSEARCH_PATH.getUrl() + "?q=*&format=xml&src=" + OPENSEARCH_SOURCE_ID;
 
-        when().get(queryUrl).then().log().all().assertThat().body(hasXPath(
-                "/metacards/metacard/string[@name='" + Metacard.TITLE + "']/value[text()='"
-                        + RECORD_TITLE_1 + "']"), hasXPath("/metacards/metacard/geometry/value"),
-                hasXPath("/metacards/metacard/string[@name='" + Metacard.TITLE + "']/value[text()='"
-                        + RECORD_TITLE_2 + "']"), hasXPath("/metacards/metacard/stringxml"));
+        when().get(queryUrl)
```
Comment (oscaritoro) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: All these "when()." lines are using RestAssured. The RestAssured formatting is different than the java standards and gets reformatted all weird by our reformatter... we should probably revert the formatting changes that occurred on the RestAssured lines and add `//@formatter:off` and `//@formatter:on` around them to prevent this from happening in the future. (I know we do this in other itests)

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java (hunk 4 above)
Comment (gordocanchola) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: Removed the unused getters.
Added proper javadoc. 
Ran formatter against the file

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/endpoint/CatalogEndpoint.java (hunk 6 above)
Comment (gordocanchola) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/endpoint/CatalogEndpoint.java]: public static final is the default for constants defined in an interface

File: catalog/opensearch/catalog-opensearch-endpoint/pom.xml (hunk 7 above)
Comment (gordocanchola) [inline catalog/opensearch/catalog-opensearch-endpoint/pom.xml]: I didn't run the new formatter against the xml files for minor updates to avoid masking my changes.

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 8 above)
Comment (gordocanchola) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I didn't run the new formatter against the xml files for minor updates to avoid masking my changes.

File: catalog/spatial/csw/spatial-csw-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 11 above)
Comment (gordocanchola) [inline catalog/spatial/csw/spatial-csw-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I didn't run the new formatter against the xml files for minor updates to avoid masking my changes.

File: catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 12 above)
Comment (gordocanchola) [inline catalog/spatial/wfs/1.0.0/spatial-wfs-v1_0_0-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I didn't run the new formatter against the xml files for minor updates to avoid masking my changes.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 5 above)
Comment (gordocanchola) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: This is showing up because of formatting changes. I agree that should work.

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 8 above)
Comment (oscaritoro) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: At a minimum then maybe just run the formatter on your section... I think on all the blueprints you touched the only changes the reformatter did for me was on your new url property line.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java (hunk 4 above)
Comment (djblue) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: I think @gordocanchola really likes `this`? :trollface:

File: catalog/opensearch/catalog-opensearch-endpoint/pom.xml (hunk 7 above)
Comment (djblue) [inline catalog/opensearch/catalog-opensearch-endpoint/pom.xml]: @gordocanchola you can do the reformatting in another commit?

Comment (djblue) [inline catalog/opensearch/catalog-opensearch-endpoint/pom.xml]: Actually, after looking at the rest of the pr, it would be really nice if all the formatting changes were in another commit.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java (hunk 4 above)
Comment (gordocanchola) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/endpoint/impl/CatalogEndpointImpl.java]: Removed this for Chris

File: catalog/opensearch/catalog-opensearch-endpoint/pom.xml (hunk 7 above)
Comment (gordocanchola) [inline catalog/opensearch/catalog-opensearch-endpoint/pom.xml]: Cool. I'll push the formatting changes in another commit.

File: catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+import static org.junit.Assert.assertEquals;
+
+import java.util.Map;
+
+import org.junit.Test;
+
+public class CatalogEndpointImplTest {
+
+    @Test
+    public void testSettingEndpointPropertiesIndividually() {
```
Comment (oscaritoro) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java]: You're only testing the non-null paths here... but each setter also has logic for dealing with null that should theoretically be covered by your tests.

Comment (oscaritoro) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java]: You don't have a test that covers the usage of `setEndpointProperties()`... should you?

File: catalog/registry/pom.xml (hunk 10 above)
Comment (oscaritoro) [inline catalog/registry/pom.xml]: This still needs to be reformatted.

File: catalog/registry/catalog-registry-cswinputtransformer/pom.xml [hunk 15]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (oscaritoro) [inline catalog/registry/catalog-registry-cswinputtransformer/pom.xml]: This needs to be reformatted

File: catalog/spatial/csw/spatial-csw-common/pom.xml [hunk 16]
```diff
@@ -42,7 +44,7 @@
                                         <limit>
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.41</minimum>
+                                            <minimum>0.40</minimum>
```
Comment (oscaritoro) [inline catalog/spatial/csw/spatial-csw-common/pom.xml]: According to a verify I just did, this value can now be set to 0.42.
Also, branches is now up to 0.20; complexity up to 0.10 and lines covered is up to 0.25.

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 17]
```diff
@@ -560,18 +623,44 @@ public void testConnectedSourceStatus() {
...
+            CatalogEndpoint endpoint = (CatalogEndpoint) bundleCtx.getService(ref);
+            String urlBindingName = endpoint.getEndpointProperties()
+                    .get("urlBindingName");
+
+            assertTrue("Catalog endpoint url binding name: '" + urlBindingName + "' is expected.",
+                    expectedEndpoints.contains(urlBindingName));
+        }
     }
 
     public void setupConnectedSources() throws IOException {
```
Comment (oscaritoro) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: While not really part of your "scope", this method uses the (internal) deprecated method `createManagedService()`... should we clean this up while we're in here?

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 5 above)
Comment (oscaritoro) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: These actually show up as warnings on intelliJ, should we take the time to clear them up while we're in there?

Comment (gordocanchola) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Cleaned them up

File: catalog/registry/pom.xml (hunk 10 above)
Comment (gordocanchola) [inline catalog/registry/pom.xml]: Formatted

File: catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java (hunk 14 above)
Comment (gordocanchola) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java]: Added unit tests for alternate paths

File: catalog/registry/catalog-registry-cswinputtransformer/pom.xml (hunk 15 above)
Comment (gordocanchola) [inline catalog/registry/catalog-registry-cswinputtransformer/pom.xml]: Formatted

File: catalog/spatial/csw/spatial-csw-common/pom.xml (hunk 16 above)
Comment (gordocanchola) [inline catalog/spatial/csw/spatial-csw-common/pom.xml]: Updated the coverage numbers

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java (hunk 17 above)
Comment (gordocanchola) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: No longer using deprecated stuff

File: catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+        catalogEndpoint.setDescription(justAnyString);
+        catalogEndpoint.setId(justAnyString);
+        catalogEndpoint.setName(justAnyString);
+        catalogEndpoint.setUrl(justAnyString);
+        catalogEndpoint.setUrlBindingName(justAnyString);
+        catalogEndpoint.setVersion(justAnyString);
+
+        Map<String, String> endpointProperties = catalogEndpoint.getEndpointProperties();
+        assertFalse("Endpoint Properties empty? Hopefully not.", endpointProperties.isEmpty());
+
```
Comment (vinamartin) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java]: Should you be asserting that the properties were actually set? Or is it fine not to test that?

Comment (oscaritoro) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java]: @vinamartin I believe he already did that in another test, the point of this test is to check that they get removed when set to Null... if he were to add that here then he might as well just "merge" the two tests into one (by moving the sets to null and assert to the other test) or maybe by moving the "catalogEndpoint.set..." code to setup and just doing the asserts that they are there in one test, and setting to null/assert empty in this one.

Comment (gordocanchola) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/endpoint/impl/CatalogEndpointImplTest.java]: Added the checks for the values being set

**Comment by oscaritoro** [comment]
I'm happy now :+1:
